src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender2.cs

```diff
@@ -0,0 +1,464 @@
+                discoveryCriteria,
+                this.protocolVersion);
+            this.channel.Send(message);
+        }
+
+        #endregion
+
+        #region Execution Protocol
+
+        /// <inheritdoc />
+        public void InitializeExecution(IEnumerable<string> pathToAdditionalExtensions, bool loadOnlyWellKnownExtensions)
+        {
+            var message = this.dataSerializer.SerializePayload(
+                MessageType.ExecutionInitialize,
+                pathToAdditionalExtensions);
+            this.channel.Send(message);
+        }
+
+        /// <inheritdoc />
+        public void StartTestRun(TestRunCriteriaWithSources runCriteria, ITestRunEventsHandler eventHandler)
+        {
+            this.messageEventHandler = eventHandler;
+            this.onDisconnected = (disconnectedEventArgs) =>
+                {
+                    this.OnTestRunAbort(eventHandler, disconnectedEventArgs.Error);
+                };
+            this.onMessageReceived = (sender, args) => this.OnExecutionMessageReceived(sender, args, eventHandler);
+            this.channel.MessageReceived += this.onMessageReceived;
+
+            var message = this.dataSerializer.SerializePayload(
+                MessageType.StartTestExecutionWithSources,
+                runCriteria,
+                this.protocolVersion);
+            this.channel.Send(message);
+        }
+
+        /// <inheritdoc />
+        public void StartTestRun(TestRunCriteriaWithTests runCriteria, ITestRunEventsHandler eventHandler)
+        {
+            this.messageEventHandler = eventHandler;
+            this.onDisconnected = (disconnectedEventArgs) =>
+                {
+                    this.OnTestRunAbort(eventHandler, disconnectedEventArgs.Error);
+                };
+            this.onMessageReceived = (sender, args) => this.OnExecutionMessageReceived(sender, args, eventHandler);
+            this.channel.MessageReceived += this.onMessageReceived;
+
+            var message = this.dataSerializer.SerializePayload(
+                MessageType.StartTestExecutionWithTests,
+                runCriteria,
+                this.protocolVersion);
+            this.channel.Send(message);
+        }
+
+        /// <inheritdoc />
+        public void SendTestRunCancel()
+        {
+            this.channel?.Send(this.dataSerializer.SerializeMessage(MessageType.CancelTestRun));
+        }
+
+        /// <inheritdoc />
+        public void SendTestRunAbort()
+        {
+            this.channel?.Send(this.dataSerializer.SerializeMessage(MessageType.AbortTestRun));
+        }
+
+        #endregion
+
+        /// <inheritdoc />
+        public void EndSession()
+        {
+            if (!this.IsOperationComplete())
+            {
+                this.channel.Send(this.dataSerializer.SerializeMessage(MessageType.SessionEnd));
+            }
+        }
+
+        /// <inheritdoc />
+        public void OnClientProcessExit(string stdError)
```

Comment:
Implementation of `ITestRequestSender.OnClientProcessExit`, used to notify if test host has exited by callers.